File: apps/mobile/modules/ricecal-widgets/android/src/main/java/expo/modules/ricecalwidgets/WidgetSnapshot.kt
package expo.modules.ricecalwidgets

import org.json.JSONObject

/**
 * The app's `WidgetSnapshot`, as Kotlin reads it.
 *
 * Kept in step with `modules/ricecal-widgets/src/types.ts` by hand, exactly as
 * fragile as the Swift copy — so the failure is made loud in the same way.
 * `parse` returns null for anything it cannot read whole, and a null snapshot
 * draws the "open RiceCal" placeholder. A field renamed on one side produces six
 * blank widgets, not six widgets confidently drawing zero.
 *
 * `org.json` rather than a parser with a dependency behind it. This runs inside
 * a broadcast receiver with a few milliseconds to live, and the document is
 * twenty fields.
 */
data class WidgetSnapshot(
  val version: Int,
  val updatedAt: Long,
  val date: String,
  val theme: String,
  val hasBudget: Boolean,
  val kcal: Kcal,
  val macros: Macros,
  val water: Water,
  val weight: Weight?,
  val entries: List<Entry>,
) {
  data class Kcal(
    val left: String,
    val eaten: String,
    val budget: String,
    val fraction: Double,
    val over: Boolean,
  )

  /** The grams, "182g". The macro's NAME is the widget's own word for it. */
  data class Bar(val fraction: Double, val label: String)

  data class Macros(val carbs: Bar, val protein: Bar, val fat: Bar)

  data class Water(
    val ml: Int,
    val goalMl: Int,
    val label: String,
    val goalLabel: String,
    val fraction: Double,
  )

  data class Weight(
    val value: String,
    val unit: String,
    val change: String,
    /**
     * Whether the change is a gain, which is what colours the pill.
     *
     * READ WITH A DEFAULT rather than required, because the store outlives the
     * build that wrote it: a phone that updates the app overnight renders the
     * document the previous build left until the app is next opened, and a
     * required field added after that document was written would fail the whole
     * parse and blank six widgets. False reads as "not a gain", which is the
     * colour this pill had before the field existed.
     */
    val up: Boolean,
    val weeks: List<Double>,
  )

  data class Entry(val name: String, val kcal: String)

  companion object {
    fun parse(json: String): WidgetSnapshot? = runCatching {
      val root = JSONObject(json)
      val kcal = root.getJSONObject("kcal")
      val macros = root.getJSONObject("macros")
      val water = root.getJSONObject("water")

      WidgetSnapshot(
        version = root.getInt("version"),
        updatedAt = root.getLong("updatedAt"),
        date = root.getString("date"),
        theme = root.getString("theme"),
        hasBudget = root.getBoolean("hasBudget"),
        kcal = Kcal(
          left = kcal.getString("left"),
          eaten = kcal.getString("eaten"),
          budget = kcal.getString("budget"),
          fraction = kcal.getDouble("fraction"),
          over = kcal.getBoolean("over"),
        ),
        macros = Macros(
          carbs = bar(macros.getJSONObject("carbs")),
          protein = bar(macros.getJSONObject("protein")),
          fat = bar(macros.getJSONObject("fat")),
        ),
        water = Water(
          ml = water.getInt("ml"),
          goalMl = water.getInt("goalMl"),
          label = water.getString("label"),
          goalLabel = water.getString("goalLabel"),
          fraction = water.getDouble("fraction"),
        ),
        // Genuinely absent on an account nobody has ever weighed, which is not
        // a parse failure and must not be treated as one.
        weight = root.optJSONObject("weight")?.let {
          val weeks = it.getJSONArray("weeks")
          Weight(
            value = it.getString("value"),
            unit = it.getString("unit"),
            change = it.getString("change"),
            up = it.optBoolean("up", false),
            weeks = (0 until weeks.length()).map { index -> weeks.getDouble(index) },
          )
        },
        entries = root.getJSONArray("entries").let { array ->
          (0 until array.length()).map { index ->
            val entry = array.getJSONObject(index)
            Entry(name = entry.getString("name"), kcal = entry.getString("kcal"))
          }
        },
      )
    }.getOrNull()

    private fun bar(json: JSONObject) =
      Bar(fraction = json.getDouble("fraction"), label = json.getString("label"))
  }

  /**
   * `theme` is deliberately not read on Android. On iOS a WidgetKit view resolves
   * its colours in Swift and can be handed the other palette; here a widget is a
   * `RemoteViews` tree inflated by the launcher, and its colours come from
   * `values/` and `values-night/`, picked by the launcher's configuration before
   * this code runs.
   *
   * Overriding that would mean setting every colour programmatically, which
   * `RemoteViews` can only do for text (`setBackgroundTintList` is API 31 and
   * this app's floor is 26), or shipping a second copy of every drawable.
   * Following the launcher's theme is what every other Android widget does.
   */
}

File: apps/mobile/widgets/ios/WidgetSnapshot.swift
import Foundation

/**
 The app's `WidgetSnapshot`, as Swift reads it.

 Kept in step with `modules/ricecal-widgets/src/types.ts` by hand, which is
 exactly as fragile as it sounds — so the failure is made loud rather than
 quiet. `Codable` throws on a missing key, `decode` below turns that into
 `nil`, and every widget draws its "open RiceCal" placeholder. A field renamed
 on one side produces six blank widgets, not six widgets confidently drawing
 zero.

 Optional fields are the ones that are genuinely absent rather than the ones
 that might be: `weight` is null on an account nobody has ever weighed.
 */
struct WidgetSnapshot: Codable {
  let version: Int
  let updatedAt: Double
  let date: String
  let theme: String
  let hasBudget: Bool
  let kcal: Kcal
  let macros: Macros
  let water: Water
  let weight: Weight?
  let entries: [Entry]

  struct Kcal: Codable {
    let left: String
    let eaten: String
    let budget: String
    let fraction: Double
    let over: Bool
  }

  struct Bar: Codable {
    let fraction: Double
    /// The grams, "182g". The macro's NAME is the widget's own word for it.
    let label: String
  }

  struct Macros: Codable {
    let carbs: Bar
    let protein: Bar
    let fat: Bar
  }

  struct Water: Codable {
    let ml: Int
    let goalMl: Int
    let label: String
    let goalLabel: String
    let fraction: Double
  }

  struct Weight: Codable {
    let value: String
    let unit: String
    let change: String
    /**
     Whether the change is a gain, which is what colours the pill.

     OPTIONAL BECAUSE THE STORE OUTLIVES THE BUILD THAT WROTE IT. `Codable`
     throws on a missing key and `load` turns a throw into six blank widgets,
     so a field added after a snapshot was written has to be allowed to be
     absent — a phone that updates the app overnight renders the old document
     until the app is next opened. Missing reads as "not a gain", which is the
     colour this pill had before the field existed.
     */
    let up: Bool?
    let weeks: [Double]
  }

  struct Entry: Codable {
    let name: String
    let kcal: String
  }

  /**
   How many bars the weight chart is laid out for.

   The app sends between one and eight weekly averages, and the chart divides
   its width by THIS rather than by however many arrived — so a short history
   is narrow bars at the left rather than a few fat ones filling the card.
   Must match `WEEKS` in `features/widgets/snapshot.ts`.
   */
  static let weeks = 8

  /// The only version this build understands. See `version` in `types.ts`.
  static let supportedVersion = 1

  /**
   Read what the app published, or nothing.

   Three ways to get nothing, and they are deliberately the same answer: no
   snapshot yet (a fresh install), a snapshot this build is too old to read, and
   a snapshot describing a day that has since ended. The third is the one worth
   naming — a phone left on the bedside table would otherwise present last
   night's total as this morning's, and there is no timeline entry that can
   correct it because the widget cannot recompute anything.
   */
  static func load(today: String = WidgetSnapshot.todayKey()) -> WidgetSnapshot? {
    guard let json = RiceCalWidgetStore.snapshotJSON(),
      let data = json.data(using: .utf8),
      let snapshot = try? JSONDecoder().decode(WidgetSnapshot.self, from: data),
      snapshot.version == supportedVersion,
      snapshot.date == today
    else { return nil }
    return snapshot
  }

  /// `yyyy-MM-dd` in the phone's own zone, which is the key the app writes.
  static func todayKey(_ now: Date = Date()) -> String {
    let formatter = DateFormatter()
    formatter.calendar = Calendar(identifier: .gregorian)
    formatter.locale = Locale(identifier: "en_US_POSIX")
    formatter.dateFormat = "yyyy-MM-dd"
    return formatter.string(from: now)
  }

  /// Whether to draw the dark palette, given what the phone is doing.
  func prefersDark(system: Bool) -> Bool {
    switch theme {
    case "dark": return true
    case "light": return false
    default: return system
    }
  }
}
